### ISSUE TITLE
Can't fetch all private flavors in a single request

### ISSUE THREAD

**Issue by mfranc213** [gophercloud/gophercloud#222]
Gophercloud's Flavor struct is missing an `IsPublic` field.  Compare:
	[Nova's Flavor object](https://github.com/openstack/nova/blob/stable/newton/nova/objects/flavor.py#L221) with [Gophercloud's Flavor struct](https://github.com/gophercloud/gophercloud/blob/magnum/openstack/compute/v2/flavors/requests.go#L17). Furthermore, Gophercloud's *flavors* requests don't allow an `is_public` to be specified.
	
In *flavors* requests made to Nova's API, the `is_public` argument is optional. When the argument is not supplied, Nova uses [the value](https://github.com/openstack/nova/blob/stable/newton/nova/api/openstack/compute/flavors.py#L64) **True**. The argument is also [ternary](https://github.com/openstack/python-novaclient/blob/stable/newton/novaclient/v2/flavors.py#L112).  When the argument is supplied, the following values are allowed:

* **True** returns private flavors associated with that project as well as (a bit confusingly) public flavors
* **False** (admin only) returns only private flavors, across all projects
* **None** (admin only) is interpreted by Nova as a request to [not filter](https://github.com/openstack/nova/blob/stable/newton/nova/api/openstack/compute/flavors.py#L67) by `is_public`, so it returns all public and private flavors across all projects.

All *four* options are tested in novaclient: [no is_public argument](https://github.com/openstack/python-novaclient/blob/stable/newton/novaclient/tests/unit/v2/test_flavors.py#L35), [None](https://github.com/openstack/python-novaclient/blob/stable/newton/novaclient/tests/unit/v2/test_flavors.py#L60), [False](https://github.com/openstack/python-novaclient/blob/stable/newton/novaclient/tests/unit/v2/test_flavors.py#L67), and [True](https://github.com/openstack/python-novaclient/blob/stable/newton/novaclient/tests/unit/v2/test_flavors.py#L74).  **True** is the default value in novaclient's [method signature](https://github.com/openstack/python-novaclient/blob/stable/newton/novaclient/v2/flavors.py#L95). 

Because Gophercloud's flavor requests don't allow `is_public` to be specified, Nova uses the value **True**.  So there is no way, currently, to obtain the list of all private flavors other than by fetching flavors project by project, but even then, because Gophercloud's Flavor struct doesn't contain the IsPublic field, there is no way to distinguish public from private flavors through Gophercloud.

**Comment by mfranc213** [comment]
I have a fix ready and am trying to push it to https://github.com/mfranc213/gophercloud ...

**Comment by mfranc213** [comment]
Hi Joe,

Thanks for the thumbs-up.  Are you a core?  And, if so, can I create the PR?  (And, if not, is there a list of cores?)

Margaret

**Comment by jtopjian** [comment]
I'm not core, but your analysis looks correct to me. I would say go ahead and create the PR. 😄 

There are a few other places where the ternary true/false/none has come into play. Off the top of my head, specifying a subnet gateway is one of them. You can see how it was handled here: https://github.com/gophercloud/gophercloud/commit/19e713b71ea0ba4c56057127841ad792fe249782

**Comment by mfranc213** [comment]
Thank you, Joe.

**Comment by jtopjian** [comment]
No problem.

At first I wondered why you set `IsPublic` to string in `ListOpts`, but then I saw that `is_public=None` is a completely valid request to Nova. :)

Thinking out loud: If `true`, `false`, and `None` are the only three valid options, another solution is to create a type, like is done [here](https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/extensions/lbaas_v2/listeners/requests.go). So you'd have something like:

```go
type FlavorType string
const (
        Public FlavorType = "true"
        Private FlavorType = "false"
        None FlavorType = "None"
)

type ListOpts struct {
        ...
        IsPublic FlavorType `q:is_public`
}
```

The benefit of this would be to prevent someone from passing `IsPublic: "foobar"`.

**Comment by mfranc213** [comment]
Thank you, Joe.  I was worried about that exact problem and wasn't sure how to address it as I'm relatively new to Go.  I will take make the change you propose.

**Comment by jrperritt** [comment]
closed by #223
